File: src-tauri/src/commands/report.rs
use crate::contexts::report::{ReportData, ReportPattern, ReportPatternParams, ReportTemplate};
use crate::error::CommandError;
use crate::lifecycle::app_state::AppState;
use tauri::State;

#[tauri::command]
#[specta::specta]
pub async fn list_report_patterns(
    state: State<'_, AppState>,
) -> Result<Vec<ReportPattern>, CommandError> {
    state
        .report_pattern_repo
        .list_patterns()
        .await
        .map_err(CommandError::from)
}

#[tauri::command]
#[specta::specta]
pub async fn create_report_pattern(
    params: ReportPatternParams,
    state: State<'_, AppState>,
) -> Result<ReportPattern, CommandError> {
    state
        .report_pattern_repo
        .create_pattern(&params)
        .await
        .map_err(CommandError::from)
}

#[tauri::command]
#[specta::specta]
pub async fn update_report_pattern(
    id: String,
    params: ReportPatternParams,
    state: State<'_, AppState>,
) -> Result<ReportPattern, CommandError> {
    state
        .report_pattern_repo
        .update_pattern(&id, &params)
        .await
        .map_err(CommandError::from)
}

#[tauri::command]
#[specta::specta]
pub async fn toggle_report_pattern(
    id: String,
    enabled: bool,
    state: State<'_, AppState>,
) -> Result<(), CommandError> {
    state
        .report_pattern_repo
        .toggle_pattern(&id, enabled)
        .await
        .map_err(CommandError::from)
}

#[tauri::command]
#[specta::specta]
pub async fn delete_report_pattern(
    id: String,
    state: State<'_, AppState>,
) -> Result<(), CommandError> {
    state
        .report_pattern_repo
        .delete_pattern(&id)
        .await
        .map_err(CommandError::from)
}

// ── Report Templates ─────────────────────────────────────────────────────────

#[tauri::command]
#[specta::specta]
pub async fn list_report_templates(
    state: State<'_, AppState>,
) -> Result<Vec<ReportTemplate>, CommandError> {
    state
        .report_template_repo
        .list_templates()
        .await
        .map_err(CommandError::from)
}

#[tauri::command]
#[specta::specta]
pub async fn get_report_template(
    id: String,
    state: State<'_, AppState>,
) -> Result<ReportTemplate, CommandError> {
    state
        .report_template_repo
        .get_template(&id)
        .await
        .map_err(CommandError::from)
}

#[tauri::command]
#[specta::specta]
pub async fn create_report_template(
    template: ReportTemplate,
    state: State<'_, AppState>,
) -> Result<(), CommandError> {
    state
        .report_template_repo
        .create_template(&template)
        .await
        .map_err(CommandError::from)
}

#[tauri::command]
#[specta::specta]
pub async fn update_report_template(
    template: ReportTemplate,
    state: State<'_, AppState>,
) -> Result<(), CommandError> {
    state
        .report_template_repo
        .update_template(&template)
        .await
        .map_err(CommandError::from)
}

#[tauri::command]
#[specta::specta]
pub async fn set_active_report_template(
    id: String,
    state: State<'_, AppState>,
) -> Result<(), CommandError> {
    state
        .report_template_repo
        .set_active_template(&id)
        .await
        .map_err(CommandError::from)
}

#[tauri::command]
#[specta::specta]
pub async fn delete_report_template(
    id: String,
    state: State<'_, AppState>,
) -> Result<(), CommandError> {
    state
        .report_template_repo
        .delete_template(&id)
        .await
        .map_err(CommandError::from)
}

#[tauri::command]
#[specta::specta]
pub async fn generate_report_data(
    job_id: String,
    state: State<'_, AppState>,
) -> Result<ReportData, CommandError> {
    state
        .report_context
        .generate_report(&job_id)
        .await
        .map_err(CommandError::from)
}

File: src/api/report.ts
import { commands } from "@/src/bindings";
import type {
  ReportPattern,
  ReportPatternParams,
  ReportData,
  ReportTemplate,
} from "@/src/bindings";

export type { ReportPattern, ReportPatternParams, ReportData, ReportTemplate };

export async function listReportPatterns(): Promise<ReportPattern[]> {
  const res = await commands.listReportPatterns();
  if (res.status === "ok") return res.data;
  throw new Error(res.error ?? "Failed to list report patterns");
}

export async function createReportPattern(params: ReportPatternParams): Promise<ReportPattern> {
  const res = await commands.createReportPattern(params);
  if (res.status === "ok") return res.data;
  throw new Error(res.error ?? "Failed to create report pattern");
}

export async function updateReportPattern(id: string, params: ReportPatternParams): Promise<ReportPattern> {
  const res = await commands.updateReportPattern(id, params);
  if (res.status === "ok") return res.data;
  throw new Error(res.error ?? "Failed to update report pattern");
}

export async function toggleReportPattern(id: string, enabled: boolean): Promise<void> {
  const res = await commands.toggleReportPattern(id, enabled);
  if (res.status !== "ok") throw new Error(res.error ?? "Failed to toggle report pattern");
}

export async function deleteReportPattern(id: string): Promise<void> {
  const res = await commands.deleteReportPattern(id);
  if (res.status !== "ok") throw new Error(res.error ?? "Failed to delete report pattern");
}

export async function generateReportData(jobId: string): Promise<ReportData> {
  const res = await commands.generateReportData(jobId);
  if (res.status === "ok") return res.data;
  throw new Error(res.error ?? "Failed to generate report data");
}

// ── Report Templates ────────────────────────────────────────────────────────

export async function listReportTemplates(): Promise<ReportTemplate[]> {
  const res = await commands.listReportTemplates();
  if (res.status === "ok") return res.data;
  throw new Error(res.error ?? "Failed to list report templates");
}

export async function getReportTemplate(id: string): Promise<ReportTemplate> {
  const res = await commands.getReportTemplate(id);
  if (res.status === "ok") return res.data;
  throw new Error(res.error ?? "Failed to get report template");
}

export async function createReportTemplate(template: ReportTemplate): Promise<void> {
  const res = await commands.createReportTemplate(template);
  if (res.status !== "ok") throw new Error(res.error ?? "Failed to create report template");
}

export async function updateReportTemplate(template: ReportTemplate): Promise<void> {
  const res = await commands.updateReportTemplate(template);
  if (res.status !== "ok") throw new Error(res.error ?? "Failed to update report template");
}

export async function setActiveReportTemplate(id: string): Promise<void> {
  const res = await commands.setActiveReportTemplate(id);
  if (res.status !== "ok") throw new Error(res.error ?? "Failed to set active template");
}

export async function deleteReportTemplate(id: string): Promise<void> {
  const res = await commands.deleteReportTemplate(id);
  if (res.status !== "ok") throw new Error(res.error ?? "Failed to delete report template");
}
